photon-ml/src/main/scala/com/linkedin/photon/ml/function/HessianDiagonalAggregator.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg._
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.function.glm.PointwiseLossFunction
+
+/**
+ * An aggregator to perform calculation of the Hessian diagonal vector. Both Iterable and RDD data share the same logic
+ * for data aggregation.
+ *
+ * This class is heavily influenced by the HessianVectorAggregator and the ValueAndGradientAggregator.
+ *
+ * @param func A single loss function for the generalized linear model
+ * @param dim The dimension (number of features) of the aggregator
+ */
+@SerialVersionUID(1L)
+protected[ml] class HessianDiagonalAggregator(func: PointwiseLossFunction, val dim: Int) extends Serializable {
```

Comment:
The original code I modified didn't use normalization for the Hessian diagonal. Do you recall why?

The diagonal is only used for computing variances, necessary only at the end of model training and not affecting any subsequent computations. At that time, the optimization problem should have the final, non-normalized coefficient means which can be used directly to compute the variances. Does that sound right?